File: Algorithms/166.FractionToRecurringDecimal.cpp
class Solution {
    public:
        string fractionToDecimal(int numerator2, int denominator2) {
            long long numerator = numerator2;
            long long denominator = denominator2;
            string result = "";
            if (numerator * denominator < 0)
                result = "-";
            numerator = abs(numerator);
            denominator = abs(denominator);
            if (numerator % denominator == 0)
                return result + to_string(numerator / denominator);
            result += to_string(numerator / denominator) + ".";
            numerator %= denominator;
            vector<int> reminders;
            vector<int> numerators;
            numerators.push_back(numerator);
            numerator *= 10;
            while (numerator) {
                reminders.push_back(numerator / denominator);
                numerator %= denominator;
                vector<int>::iterator i = find(numerators.begin(), numerators.end(), numerator);
                if (i != numerators.end()) {
                    int left = i - numerators.begin();
                    for (int t=0; t < left; t++)
                        result += to_string(reminders[t]);
                    result += "(";
                    for (; left < reminders.size(); left++)
                        result += to_string(reminders[left]);
                    result += ")";
                    return result;
                }
                numerators.push_back(numerator);
                numerator *= 10;
            }
            for (vector<int>::iterator index = reminders.begin(); index != reminders.end(); index++) {
                result += to_string(*index);
            }
            return result;
        }
};

File: Algorithms/py/166.FractionToRecurringDecimal.py
class Solution:
    # @return a string
    def fractionToDecimal(self, numerator, denominator):
        if numerator * denominator >= 0:
            result = ""
        else:
            result = "-"
        numerator = abs(numerator)
        denominator = abs(denominator)
        if numerator % denominator == 0:
            return result + str(numerator / denominator)
        left = numerator / denominator
        numerator %= denominator
        reminders = []
        numerators = [numerator]
        numerator *= 10
        while numerator:
            reminders.append(str(numerator / denominator))
            numerator %= denominator
            if numerator in numerators:
                i = numerators.index(numerator)
                return result + "%s.%s(%s)" % (left, "".join(reminders[:i]), "".join(reminders[i:]))
            numerators.append(numerator)
            numerator *= 10
        return result + "%s.%s" % (left, "".join(reminders))
